boto_addins/s3.py

```diff
@@ -189,15 +185,55 @@ def async_exist(self):
 
 
 @gen.coroutine
-def async_http_download(source, destination):
+def fetch_request(request, client=None, retry_callback=None,
+                  attempts=1, attempts_sleep=0.2):
+    if client is None:
+        client = simple_httpclient.SimpleAsyncHTTPClient()
+
+    # Fail faster on connection if we can retry
+    request.connect_timeout = 5 if attempts > 1 else 20
+
+    while attempts:
+        try:
+            resp = yield client.fetch(request)
+        except httpclient.HTTPError as e:
+            # retry on s3 errors
+            if e.code not in (500, 599):
+                raise
+        else:
+            raise gen.Return(resp)
+
+        attempts -= 1
+        if not attempts:
+            raise
+
+        if attempts_sleep:
+            yield gen.sleep(attempts_sleep)
+        if retry_callback:
+            yield retry_callback(request, attempts)
+
+
+@gen.coroutine
+def async_http_download(source, destination, attempts=3):
     tmp_name = '{}.{}'.format(destination, str(uuid4())[:8])
 
     try:
         with open(tmp_name, 'w') as iobuffer:
-            yield curl_httpclient.CurlAsyncHTTPClient().fetch(
-                source,
-                streaming_callback=iobuffer.write,  # Direct write to file.
-                request_timeout=10 * 60,  # Timeout for whole request, not tcp.
+
+            @gen.coroutine
+            def reset_iobuffer(*args):
+                iobuffer.seek(0)
```

Comment:
Tornado doesn't have API for nonblocking disk IO. Anyway, `seek/truncate` are much less expensive than `write`.